woocommerce/woocommerce-android
# Issue/3726 handle card errors

Parent issue #3726 

Merge instructions:
1. ~Make sure https://github.com/woocommerce/woocommerce-android/pull/4021 is merged~
2. ~Remove "do not merge" label~
3. Merge this PR

This PR adds support for:
1. Showing "Card declined" error on the UI
2. Showing "No internet connection" error on the UI
3. Accepting "Payment already captured" error from capturePayment endpoint as a success

- Refactors different errors in CardPaymentStatus into a single PaymentFailed with enum indicating its type.
- Introduces PaymentErrorMapper which takes care of converting different types of errors from different sources into CardPaymentStatus.PaymentFailed
- Connects the errors to the UI + fixes buggy layout of the payment dialog
- Adds an extra check to "acceptPayment" and emits an error if StripeTerminal is not initialized

Notes:
- Ignore styles on the UI -> we plan to polish the UI and ask a designer for review when we are done with everything else
- I realize this PR is quite big, but the changes are very related (except of two lines). Also a lot of the changes are newly added unit tests or layout changes which don't need to be reviewed so I hope it won't be too complex to review. However, let me know if you'd prefer me to somehow split it.

To Test:
- Use the simulated reader as it's not possible to simulate payments errors using a hw reader
- Your testing site needs to have WCPay in dev mode installed and activated
1. Open App settings
2. Tap on Manage card reader
3. Check "Use simulated reader" checkbox
4. Tap on "Connect to reader" button
5. Go back and open a detail of an order which is in US dollars
6. Connect debugger (you can still use CI build)
7. Add breakpoint to [this line](https://github.com/woocommerce/woocommerce-android/blob/9738166d32da1b82637f0461d4fea46bc1a0613b/libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt#L61). Make it non-suspandable (uncheck "suspend" checkbox) and check "Evaluate and log" and enter `amountInSmallestCurrencyUnit = 101`. <img width="1063" alt="Screenshot 2021-05-12 at 15 11 16" src="https://user-images.githubusercontent.com/2261188/118038130-59ba8780-b334-11eb-8e80-bd318e10ac6d.png">
8. Tap on "Collect payment" button
9. Notice the payment fails with "Card declined" error
10. Disable the breakpoint and cancel the payment flow (dismiss the dialog)
11. Turn on the airplane mode
12. Tap on "Collect payment"
13. Notice the payment fails with "no network" error
14. Turn off the airplane mode
15. Tap on "Retry"
16. Notice the payment succeeds

Update release notes:
